Fix Tab-Group Header not displayed

gerteck opened the pull request:
**What is the purpose of this pull request?**

- [ ] Documentation update
- [x] Bug fix
- [ ] Feature addition or enhancement
- [x] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

Fixes #2524

**Overview of changes:**

Header does not show up for the `TabGroup` component. This was likely due to the header being passed as a prop instead of a named slot into `Dropdown`. ~To fix this, pass header as named slot.~ To fix this, the header is now treated as a prop.

**Anything you'd like to highlight/discuss:**

From what I understand:
* `TabGroup` internally consists of the dropdown tab component. However, internally, TabGroup does not render this dropdown, but instead adds itself to the parent `TabSet`, which is then in charge of rendering children `Tab` or `TabGroup`. 
* In `TabSet`, the header was originally passed into Dropdown from Tabset as an attribute / vue prop of `Dropdown`. `Dropdown` then uses this prop to display the header:
  * [commit history](https://github.com/MarkBind/vue-strap/commit/96adcfb15cc90d7ccba796ab5771267eccfdef9a?diff=split&w=0)
* After that, to add markdown support to the header, the header was passed using slots. I would like to see if this is the source of the bug, but at this point, I can't seem to test this version of MarkBind, due to depreciated dependencies/errors faced.
  * [commit history](https://github.com/MarkBind/vue-strap/commit/04dd7e210f6cabea02c0e6e766869a1b1ea155ab) 

**[ Requesting Comments ]**: I don't quite understand how the Vue component's prop is being passed as a named slot. Referring to the [developer guide](https://markbind.org/devdocs/devGuide/development/writingComponents.html), I tried to trace the processing but I think I might be facing a skill issue. I'm confused as it says that "MarkBind attributes are passed to the Vue component as props. The type of the prop will be a String." However, the prop is able to be processed as a named slot.

This seems to be the core of the issue behind the bug. In `Tabset`, the header is being passed into `dropdown` similarly as a prop, which is not registered as a namedSlot, and ends up as an attribute when rendered. ~To fix this, I directly passed the header as a named slot. However the discrepancy and origin of the bug leaves me uneasy.~

**Illustration**

Sample Code in MarkBind Syntax (where header is passed as an attribute):
```
<!-- header is being passed as a prop -->
<dropdown header="Prop Header" class="w-100"> 
    <ul slot="dropdown-menu" class="dropdown-menu">
      <li><a href="#dropdown" class="dropdown-item">Action</a></li>
      <li><a href="#dropdown" class="dropdown-item">Separated link</a></li>
    </ul>
  </dropdown>
```

Dropdown Vue Component Template (where header is accepted as a named slot):

https://github.com/MarkBind/markbind/blob/2c8276468592f964663ff3193e57345b63fa7f70/packages/vue-components/src/Dropdown.vue#L28-L51

Output:

While the header was passed in as a prop, it is parsed as a named slot.

![image](https://github.com/MarkBind/markbind/assets/111064611/6aa8868a-80d1-49bf-af99-f97b6f745b20)



Some references:
* [Legacy VueStrap Dropdown](https://github.com/yuche/vue-strap/blob/master/src/Dropdown.vue)
* [Add Embedded HTML in dropdown Commit](https://github.com/MarkBind/vue-strap/commit/e9c4ec9b8966ba1ea7032b9294bb68651527e804)
* [Update of header slots for Vue Components Commit](https://github.com/MarkBind/markbind/commit/640d18be106b5ca8b0b6ab0a60a21b3d19139b67)


**Testing instructions:**

<!--
  Any special testing instructions **not including** our automated tests.
-->

Refer to user guide: https://markbind.org/userGuide/components/presentation.html#tabs
Refer to PR's localhost user guide: http://127.0.0.1:8080/userGuide/components/presentation.html#tabs


Before:
![image](https://github.com/gerteck/markbind/assets/111064611/2695a4ea-a77a-4835-a806-5e92f337e2a9)

After:
![image](https://github.com/gerteck/markbind/assets/111064611/8d0d4266-851e-466d-890d-4249c9160e4a)


**Proposed commit message: (wrap lines at 72 characters)**

<!--
  See this link for more info on how to write a good commit message:
  https://se-education.org/guides/conventions/git.html

  As best as possible, write a succinct commit title in 50 characters

|---------This is the width of 50 chars----------|
|-----------This is the width of 72 chars for your reference-----------|
-->

Fix TabGroup Header Rendering Issue

The header for the TabGroup component was not displaying 
correctly, due to the header being passed as a prop but 
treated as a named slot in the Dropdown component. 
To fix this, the header is now treated as a prop.


---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [ ] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes <!-- It's tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

<!--
  We'll try our best to get to your PR within a week.
  If we haven't gotten to it then, or if your pull request resolves an urgent item, feel free to give us a ping!
-->

---

**Reviewer checklist:**

Indicate the [SEMVER](https://semver.org/) impact of the PR:
- [ ] Major (when you make incompatible API changes)
- [ ] Minor (when you add functionality in a backward compatible manner)
- [ ] Patch (when you make backward compatible bug fixes)

At the end of the review, please label the PR with the appropriate label: `r.Major`, `r.Minor`, `r.Patch`.

Breaking change release note preparation (if applicable):
- To be included in the release note for any feature that is made obsolete/breaking

> Give a brief explanation note about:
>   - what was the old feature that was made obsolete
>   - any replacement feature (if any), and
>   - how the author should modify his website to migrate from the old feature to the replacement feature (if possible).

lhw-1 commented:
> I don't quite understand how the Vue component's prop is being passed as a named slot. Referring to the [developer guide](https://markbind.org/devdocs/devGuide/development/writingComponents.html), I tried to trace the processing but I think I might be facing a skill issue. I'm confused as it says that "MarkBind attributes are passed to the Vue component as props. The type of the prop will be a String." However, the prop is able to be processed as a named slot.

The [`createSlotTemplateNode`](https://github.com/MarkBind/markbind/blob/master/packages/core/src/html/elements.ts#L11-L13) method is used to convert slots into template nodes directly; this is called by the [`processAttributeWithoutOverride`](https://github.com/MarkBind/markbind/blob/master/packages/core/src/html/MdAttributeRenderer.ts#L30-L50) method, which is in turned called for specified named slots that we are looking out for. This is how we process named slots - hopefully this addresses your question (and if it doesn't, do let me know).

Regardless, this issue might need further investigation 😓

lhw-1 commented on packages/vue-components/src/Tabset.vue:
```diff
@@ -24,10 +24,12 @@
           :class="{active:t.active}"
           :disabled="t.disabled"
         >
+          <template #header>
+            <span v-html="t.headerRendered"></span>
+          </template>
```
Instead of making this change (since we rely on the conversions from slots to templates, and do not use header templates directly in our `.vue` files), a suggestion I have is to treat the `header` as a prop instead?

In `Dropdown.vue`, we can first include the `header` as a prop:

```
props: {
  ...
  header: {
    type: String,
    default: '',
  },
  ...
```

And then we can replace the `<slot name="header"></slot>` under the `v-if`, with `{{ header }}`. We could also rename `header` to something like `tabgroup-header` to differentiate it as well.

This might be a less hacky solution, but if we can investigate the issue deeper and find a better solution, that would be best 👍

gerteck commented:
>
> The [`createSlotTemplateNode`](https://github.com/MarkBind/markbind/blob/master/packages/core/src/html/elements.ts#L11-L13) method is used to convert slots into template nodes directly; this is called by the [`processAttributeWithoutOverride`](https://github.com/MarkBind/markbind/blob/master/packages/core/src/html/MdAttributeRenderer.ts#L30-L50) method, which is in turned called for specified named slots that we are looking out for. This is how we process named slots - hopefully this addresses your question (and if it doesn't, do let me know).
> 
> Regardless, this issue might need further investigation 😓

It does address my question, thank you! Running it in the debugger with the appropriate breakpoint, I can see it too:

![image](https://github.com/MarkBind/markbind/assets/111064611/a5399c69-d2fa-4cd2-babc-42cb564987bd)

Could the reason be because `dropdown` is directly being imported and referenced by `Tabset.vue`, and consequently is not subjected to the node processing flow, hence it would be wrong to pass the slot as an attribute within the `Tabset.vue` file?

If this is the case, I agree with your suggestion to treat header as a prop which was done previously as well. It definitely does seem like a better alternative that fits the bigger picture. 😄

Tim-Siu commented:
> > The [`createSlotTemplateNode`](https://github.com/MarkBind/markbind/blob/master/packages/core/src/html/elements.ts#L11-L13) method is used to convert slots into template nodes directly; this is called by the [`processAttributeWithoutOverride`](https://github.com/MarkBind/markbind/blob/master/packages/core/src/html/MdAttributeRenderer.ts#L30-L50) method, which is in turned called for specified named slots that we are looking out for. This is how we process named slots - hopefully this addresses your question (and if it doesn't, do let me know).
> > Regardless, this issue might need further investigation 😓
> 
> It does address my question, thank you! Running it in the debugger with the appropriate breakpoint, I can see it too:
> 
> ![image](https://private-user-images.githubusercontent.com/111064611/341267323-a5399c69-d2fa-4cd2-babc-42cb564987bd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkwNDkzMjAsIm5iZiI6MTcxOTA0OTAyMCwicGF0aCI6Ii8xMTEwNjQ2MTEvMzQxMjY3MzIzLWE1Mzk5YzY5LWQyZmEtNGNkMi1iYWJjLTQyY2I1NjQ5ODdiZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyMlQwOTM3MDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jZDMzZTk1NDBkOTVlZDBhOWNmZDliNGZhNzQ3OGI1YjgwNjJhZDU3NDE0YTgwZmNjYWYxYzg3M2Y0ZWJkMGZlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Hdktoiha7kiYQmbG6oxDwo4lxskC19dFpHhIIIGw4Yc)
> 
> Am I right to say that because `dropdown` is directly being imported and referenced by `Tabset.vue`, and consequently is not subjected to the node processing flow, hence it would be wrong to pass the slot as an attribute within the `Tabset.vue` file? (As you mentioned, that we do not use header templates directly in our `.vue` files)
> 
> If this is the case, I agree with your suggestion to treat header as a prop which was done previously as well. It definitely does seem like a better alternative that fits the bigger picture. 😄

Hi @gerteck ,

Thank you for your great research!

I concur with your analysis regarding the root cause of this issue. It appears that the nodes are not subject to recursive processing, and the `<dropdown>` within `Tabset.vue` does not undergo transformation by the `NodeProcessor` to encapsulate the attribute within a `<template>` wrapper for slot content.

For the immediate solution, I support the two approaches you've outlined:
- Manually setting the slot content when importing `<dropdown>`
- Passing the header as a prop

@lhw-1's suggestions appear reasonable, and I recommend adhering to the second approach.

For a more comprehensive solution that might involve modifications to the `NodeProcessor` (such as recursively processing nodes), we can create a separate issue to address this at a later time when it becomes a priority.

yucheng11122017 commented on packages/vue-components/src/Dropdown.vue:
```diff
@@ -11,7 +11,7 @@
         data-bs-toggle="dropdown"
       >
-        <slot name="header"></slot>
+        <span v-html="tabgroupHeader"></span>
```
Hmm I'm not keen on this naming since this is only true for tab groups but not for other headers. 
Could we keep the header name?

gerteck replied:
Sure! The rationale was give greater differentiation in the way the header is passed as a prop instead of relying on slot conversion, but it is evident from the code directly. 

I would be down to try writing some test cases for the tabs! Test cases would also prevent any further regressions until a better way to pass down the header is found.

gerteck replied:
I just realised that setting it as the same name header causes two of the current tests to fail under `dropdown` as the header no longer shows up as an attribute in the html rendered. This is probably because of the slot name and the attribute prop name having the same name.

I could edit the testcases to remove the header, or simply revert to using a separate name again to prevent this issue

yucheng11122017 reviewed (commented):
Hi @gerteck thanks for the work! 
I realised that tabs doesnt have any test cases.. which is why we probably missed out on this regression in behaviour. If you have the time and capacity, you could try writing test cases for tabs. If you don't though, it's ok. Let us know and we will open a issue for this.

Tim-Siu commented:
> Hi @gerteck thanks for the work! I realised that tabs doesnt have any test cases.. which is why we probably missed out on this regression in behaviour. If you have the time and capacity, you could try writing test cases for tabs. If you don't though, it's ok. Let us know and we will open a issue for this.

I think we do have some unit test regrading tab-group in `NodeProcessor.test.ts` but more advanced unit tests or functional test may be necessary to test the intended behavior for the header.

gerteck commented:
* Added some snapshot tests for Tabset (Tabs), Tabgroup, Tabs to catch visual regressions during future changes.  
* Renamed the `header` prop back into `tabGroupHeader`, as currently, it is a workaround used only to generate tabgroups in `Tabset`, and I think it would be confusing to have both a `header` named slot and prop.
  * It also seemed to cause some regression which I still don't understand the cause of. Having an added "header" prop seems to modify the DOM of nested `dropdown`s (as in `Dropdown.spec.js`), although I don't see any visual change.


* Readded the named slot `header` 
  * I missed out a regression when I removed the named slot `header`. (The impacted use case was `navbar` with `dropdown` children in markbind syntax where it was passed as a attribute and converted into a named slot). 
  * It also looks like there are no test cases for this so I will add some as well.

Tim-Siu reviewed (approved):
Thank you for delving into adding snapshot tests for the Vue plugins. I think they are well-written and sufficient to cover the concern of the issue, i.e., Tab-Group Header not displayed in when Dropdown is nested in Tabset.

tlylt commented:
@Tim-Siu please replace the commit message with the proposed message whenever you merge the PR in the future.

Tim-Siu commented:
> @Tim-Siu please replace the commit message with the proposed message whenever you merge the PR in the future.

Sorry for the mistake. Should I update the commit message and force push?

tlylt commented:
> > @Tim-Siu please replace the commit message with the proposed message whenever you merge the PR in the future.
> 
> Sorry for the mistake. Should I update the commit message and force push?

Think we can leave it, just take note next time
